### PR TITLE
Move special functions to a separate module

### PR DESCRIPTION
**Pull request by musm**
Would be  good to separate the file to a separate folder like LinAlg to have for additional pure Julia implementations of math functions.

### CODE REVIEW

**Comment by yuyichao** [comment]
Again, please update/ping original PR instead of opening new ones!

**Comment by musm** [comment]
Not possible since I force repushed the close PR, and github doesn't let me open it back

**Comment by ViralBShah** [comment]
Why do all this moving? For now, can we leave things as they are until there is a larger overhaul/reason to move things?

**Comment by simonbyrne** [comment]
I can see the appeal: it does follow the pattern of other submodules whereby we put all the necessary files under 1 directory. Especially since `special` is already a directory.

Alternatively, there is #8536, but not much has happened on that front, and no concrete plan has emerged.

**Comment by tkelman** [comment]
There is a plan. It's https://github.com/JuliaLang/julia/issues/18795.

**Comment by simonbyrne** [comment]
It my be fair to call it a vague plan, but it is far from concrete.

**Comment by tkelman** [comment]
What's not concrete about it?

**Comment by musm** [comment]
What exactly is the plan. Is this module being pulled out of the Julia repo to its own repo for 0.6?

Also it's bothersome that  replutil.jl depends on the filename and location: https://github.com/musm/julia/blob/98466be678d1b5ad53de3deda8601462e9231f39/base/replutil.jl#L217
this will need to decoupled if the modules are moving out of the main julia repo.

**Comment by tkelman** [comment]
Special functions could be if someone implements it. I don't think the rest of math.jl should move as soon though as that's more widely used.

**Comment by tkelman** [comment]
some of these file locations are also listed as exceptions in contrib/add_license_to_files.jl

**Comment by simonbyrne** [comment]
In that case, why don't we move the special functions to a new module, so that they're ready to be split out?

**Comment by musm** [comment]
What happens in the future for packages that need a function like `erf` ? Will they all need to import the SpecialFunctions package?

IMO it might be worthy to keep some special functions in base mlib such as `erf` and gamma because of their fundamental importance in engineering and stats, etc. This practice is not uncommon to other mlibs, since they typically  guarantee a high quality implementation for such critical functions 

Thoughts ?

@simonbyrne

**Comment by ViralBShah** [comment]
Yes, we will still want some of these in base. Also, I don't see why we can't reorganize things in base until default packages happens, if there is a good reason to do them.

**Comment by tkelman** [comment]
airy, bessel, and probably dawson would be less disruptive to move in an earlier phase than erf and gamma.

**Comment by simonbyrne** [comment]
Agreed: lets move those first; others we can evaluate on a case-by-case basis. 

I think the main steps would be (until it's clear what form the stdlib would take):
 - move `special/bessel.jl` to `specialfunctions/SpecialFunctions.jl`, which contains a module of the same name; can optionally split this out into smaller files
 - all other files under `special` get moved to under `math` as in this PR.

Does that seem reasonable?

**Comment by musm** [comment]
IMO I would keep `bessel`, `gamma`, and `erf` in Math (which admittedly is most of the function in special) , and move the rest to `SpecialFunctions`. 


In case, someone has not seen this: http://www.evanmiller.org/statistical-shortcomings-in-standard-math-libraries.html

**Comment by simonbyrne** [comment]
How is that different than your current PR?

**Comment by simonbyrne** [comment]
Bessel & Airy functions have long been marked for removal (#8536): they are the messiest to call (see all the glue code, global buffer variables and various issues about Float32, Float64 and BigFloat). Why would you want to keep them in Base?

**Comment by musm** [comment]
I see. Seems a little arbitrary then to keep `erf` and `gamma` in `Math` then, no? 
So the options are:
1. Move all special functions to SpecialFunctions.jl
2. Move only erf and gamma (pretty arbitrary, but w/e)
3. Move none.

**Comment by simonbyrne** [comment]
`erf`,`erfc` and `gamma` are defined in the C math library, so there is some precedence here. Furthermore, they're also single argument functions, and used in a lot of domains. 

There is some discussion of which functions to be moved also in #4301: perhaps best to continue this there.

**Comment by ViralBShah** [comment]
It would be nice to get this merged in some shape for 0.6, even if we can't get 100% on module names.

**Comment by musm** [comment]
Should tests and stdlib/math.md  also be split ?

**Comment by musm** [comment]
Yes perhaps we need better names for these. Perhaps `Math`  and `SpecFun` are  not the most suggestive.

File: base/sysimg.jl [hunk 1]
```diff
@@ -242,6 +235,17 @@ let SOURCE_PATH = ""
...
+# core math functions
+include("floatfuncs.jl")
+include("math/math.jl")
+importall .Math
+const (√)=sqrt
+const (∛)=cbrt
+
+# special math functions
+include("special/special.jl")
+importall .SpecFun
```
Comment (tkelman) [inline base/sysimg.jl]: SpecialFunctions ?

**Comment by musm** [comment]
Ok rebased. Are these  the names we want? `Math` for elementary functions and more core math related functions and `SpecialFunctions` for *all* the special functions? Suggestions welcome. We should try to finalize this now for https://github.com/JuliaLang/julia/issues/18795 .

**Comment by musm** [comment]
Should we hold off till 0.1 for this?

**Comment by simonbyrne** [comment]
Any reason you closed this? AFAIK we still wanted to do this.

**Comment by musm** [comment]
It got a thumbs down and the comment with a couple thumbs up https://github.com/JuliaLang/julia/pull/19560#issuecomment-266399277 
doesn't seem like people want to do this now

**Comment by tkelman** [comment]
The recent versions of this PR were better than where it started.

**Comment by musm** [comment]
K, I'll  reopen and see if there is interest in getting this in.

**Pull request closed without merging**